bucket type properties in CREATE WITH clause

hmmr opened the pull request:
RTS-865

This PR is a refurbished #1328, obsoletes #1332. Requires https://github.com/basho/riak_ql/pull/113.

KV support for setting custom table properties in an optional WITH clause of the CREATE TABLE query. This allows clients to create TS tables with custom properties via query API call.

There is also a more strict and accurate validation of property names accepted in WITH clause (as well as in json supplied to `riak-admin bucket-type create` command. Now an attempt to create a bucket type with an unknown property will result in appropriate error message.

Commit:
relax property retrieval checks in a riak_kv_console eunit test

A proplist should not be relied upon to have properties in a
particular order.

Commit:
report error messages maybe_parse_table_def as single lines

let \n be added by eventual io:format

Commit:
support setting bucket props with WITH in CREATE TABLE query

Relies on lexer/parser support in riak_ql.

Instead of {ok, #ddl_v1{}}, riak_ql_parser:parse/1 now returns
{ok, {#ddl_v1{}, Properties::proplist()}}. This introduces some
cumbersomeness but has been chosen in favour of adding a new field
and letting #ddl_v1{} carry the properties in it.

Commit:
return {1021, notfound} also in get

Commit:
fix an eunit test in riak_kv_console after SQL records reorg

Commit:
consistently refer to #ddl_v1{} as ?DDL{}

Commit:
more accurate check for valid bucket type property names

In riak_kv_wm_utils:erlify_bucket_prop, rather than relying on
list_to_existing_atom, check against a list of known property names,
via a convenience fun riak_kv_bucket:is_valid_property/1.

Commit:
shorten lines, clean up whitespace

andytill commented on src/riak_kv_bucket.erl:
```diff
@@ -43,6 +44,43 @@
 
 -export_type([props/0]).
 
+-define(VALID_PROPERTIES,
```
Some of these properties cannot be used by Time Series, or are not supported by write once.

Only `dw`, `n_val`, `pr`, `pw`, `r`, `rw` and `w` should be allowed.

macintux commented on src/riak_kv_bucket.erl:
```diff
@@ -43,6 +44,43 @@
+         "table_def"
+        ]).
+
+-spec is_valid_property(string() | binary()) -> boolean().
```
Tough call. We've not constrained bucket properties in the past, and it's unclear this is a good reason to start doing so.

On the other hand, we've certainly wanted to document all possible bucket properties to get out of the JSON-on-the-CLI business for riak-admin bucket-type create, so this work may prove valuable even if we choose not to use it.

macintux replied:
Upon further reflection, will have to reject this, at least as a global change to `erlify_bucket_props`. Buckets are a riak_core concept that applications can use even if not written by Basho, hence we don't and can't have a whitelist of properties.

hmmr replied:
Quite simply, there's a list_to_existing_atom down the line, and here's why
I attempt to collect all possible values for the bucket properties here.
On 16 Mar 2016 8:24 pm, "John Daily" notifications@github.com wrote:

> In src/riak_kv_bucket.erl
> https://github.com/basho/riak_kv/pull/1369#discussion_r56388669:
> 
> > -         "postcommit",
> > -         "pr",
> > -         "precommit",
> > -         "pw",
> > -         "r",
> > -         "rw",
> > -         "small_vclock",
> > -         "w",
> > -         "write_once",
> > -         "young_vclock",
> > -         %% TS-specific ones:
> > -         "ddl",
> > -         "table_def"
> > -        ]).
> >   +
> >   +-spec is_valid_property(string() | binary()) -> boolean().
> 
> Tough call. We've not constrained bucket properties in the past, and it's
> unclear this is a good reason to start doing so.
> 
> On the other hand, we've certainly wanted to document all possible bucket
> properties to get out of the JSON-on-the-CLI business for riak-admin
> bucket-type create, so this work may prove valuable even if we choose not
> to use it.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> https://github.com/basho/riak_kv/pull/1369/files/704efa9a70dc8148468ddf16bc74504aa6494c22#r56388669

macintux replied:
Understood, but we still need to remove all of this. Catch the badarg if you need to.

andytill commented on src/riak_kv_ts_util.erl:
```diff
@@ -186,26 +191,36 @@ maybe_parse_table_def(BucketType, Props) ->
-            "    bucket type was: " ++ binary_to_list(BucketType1) ++ "\n"
-            "     table name was: " ++ binary_to_list(BucketType2) ++ "\n"}}).
+            flat_format(
+              "Time series bucket type and table name mismatch (~s != ~s)",
```
I prefer the original message, it is very easy to compare vertical columns.

If this one is here to stay it should read, Time series table and bucket type do not match.

Commit:
better diagnostics in riak_kv_wm_utils:erlify_bucket_prop

* throw meaningful exceptions instead of badarg (on list_to_existing_atom
  failing when extracting Mod:Fun for linkfun and chash_fun props) or
  badmatch (on bad bkey for jsanon);
* allow free-form custom properties (log a message when one is created).

Commit:
catch exceptions from erlify_bucket_prop in riak_kv_wm_bucket_type

else, a malformed property in json would fly to webmachine

Commit:
error message tweak

Commit:
don't keep a registry of 'common properties'

Just log an informational message when attempt is made to add a property
which did not exist as an atom.

hmmr commented:
@borshop: retry

macintux commented:
One last bit I'd like to see before this is merged: I don't see any tests to provoke the new error tuples from `erlify_bucket_prop`: `bad_linkfun_modfun`, `bad_linkfun_key`, and `bad_chash_keyfun`. A unit test for these would be lovely.

Commit:
even better exceptions from erlify_bucket_prop, with tests

macintux commented:
:+1: df1862e

hmmr commented:
@borshop: merge

macintux commented:
@borshop merge
